File: src/Python/Preprocessing/mkf2e.py
from numpy import *
from getelemface import *
from sortrows import *

def mkf2e(t,elemtype,nd):

    ne = t.shape[1]
    face = getelemface(nd,elemtype)
    nvf = face.shape[0]
    nfe = face.shape[1]

    # sort faces for all elements to make two identical faces have the same numbering order
    N = nfe*ne;
    tf = reshape(t[face.flatten(order='F'),:], (nvf, N), order='F');
    tf = sort(tf, axis=0);

    tf, jx = sortrows(tf.T); # do this so that two identical faces are next each other
    tf = tf.T;
    dx = tf[:,1:N]-tf[:,0:(N-1)]; # do this to find two identical faces
    dx = sum(dx*dx,axis=0);           # identical faces corresponds to zero entries
    in1 = nonzero(dx.flatten('F')==0)[0];           # interior global faces connected to 1st elements (1st identical faces)
    in1 = in1.flatten('F');
    in2 = in1 + 1;                 # interior global faces connected to 2nd elements (2nd identical faces)

    for i in range(0,len(in1)):
        if jx[in1[i]] > jx[in2[i]]:
            tm = jx[in1[i]];
            jx[in1[i]] = jx[in2[i]];
            jx[in2[i]] = tm;

    in3 = concatenate([in1, in2]);
    in0 = setdiff1d(arange(0,N), unique(in3)); # boundary global faces connected to 1st elements

    nf = len(in0)+len(in1);
    f2e = zeros((4,nf)).astype(int);   # allocate memory
    e2e = -ones((nfe,ne)).astype(int); # allocate memory

    # interior faces
    e1 = int64(ceil(float64(jx[in1]+1)/float(nfe))) - 1;      # 1st elements
    l1 = jx[in1] - e1*nfe;   # 1st local faces
    e2 = int64(ceil(float64(jx[in2]+1)/float(nfe))) - 1;      # 2nd elements
    l2 = jx[in2] - e2*nfe;   # 2nd local faces
    g = arange(0,len(in1));           # indices for interior faces

    f2e[0,g] = e1+1;
    f2e[1,g] = l1+1;
    f2e[2,g] = e2+1;
    f2e[3,g] = l2+1;
    for i in range(0,len(e1)):
        e2e[l1[i],e1[i]] = e2[i];
        e2e[l2[i],e2[i]] = e1[i];

    # boundary faces
    e1 = int64(ceil(float64(jx[in0]+1)/float(nfe))) - 1;      # 1st elements
    l1 = jx[in0] - e1*nfe;   # 1st local faces
    g = arange(len(in1),nf);      # indices for boundary faces
    f2e[0,g] = e1+1;
    f2e[1,g] = l1+1;

    # print(f2e.T)
    # print(e2e.T)
    # error("here")

    return f2e, e2e

File: frontends/Matlab/Mesh/mkmesh/mkf2e.m
function f2e = mkf2e(elcon, f, t2f, perm)

np = max(perm(:));
nf = size(f,1);
ne = size(t2f,1);
[npf,nfe] = size(perm);
elcon = reshape(elcon,[npf nfe ne]);

f2e = zeros(npf,nf,2);
for i = 1:nf
    fi = f(i,end-1:end); % obtain two elements sharing the same face i      
    if fi(2)>0           % face i is an interior face                
        kf = t2f(fi,:);         % obtain neighboring faces 
        i1 = kf(1,:)==i;  % obtain the index of face i in the 1st element
        i2 = kf(2,:)==i;  % obtain the index of face i in the 2nd element                                            
        j1 = elcon(:,i1,fi(1)) - (i-1)*npf;        
        j2 = elcon(:,i2,fi(2)) - (i-1)*npf;        
        f2e(:,i,1) =  (fi(1)-1)*np+perm(j1,i1);
        f2e(:,i,2) =  (fi(2)-1)*np+perm(j2,i2);        
    else % face i is a boundary face
        kf = t2f(fi(1),:); % obtain neighboring faces 
        i1 = kf(1,:)==i;  % obtain the index of face i in the 1st element                 
        j1 = elcon(:,i1,fi(1)) - (i-1)*npf;       
        f2e(:,i,1) =  (fi(1)-1)*np+perm(j1,i1);        
        f2e(:,i,2) = f2e(:,i,1);
    end        
end 
f2e = reshape(f2e,[npf*nf 2])'; 

